### DISCUSSION TITLE
How to get diagnostic output while running the firmware

### DISCUSSION THREAD

**Discussion by RogerInHawaii** [prusa3d/Prusa-Firmware#4801] [General]
I'm modifying the Open Source Firmware for my Prusa MK3S and would really like some way to output diagnostic text while it's running. I suspect it can be done by sending the text out through the same USB port that's used for flashing the firmware and having it "caught" by the Serial Monitor of the Arduino IDE.

First of all, let me know if that seems reasonable.
Secondly, if that IS reasonable, what serial object do I use within the firmware to do that?

I tried approaches like :

    SERIAL_ECHO("It got here\n");
    Serial.println(" It got here");
    Serial1.println(" It got here");

but the compiler doesn't like any of those.

Is there a recommended way of getting diagnostic text output, somehow?

I'll point out that on my Windows machine I wasn't able to get Visual Studio Code installed on it, and had to revert to installing Ubuntu and the Linux version of Visual Studio Code which does not provide breakpoints, single-stepping, variable access, or any of the standard diagnostics tools of the full Visual Studio Code, so I'm left with trying to have the firmware do some printf (or similar) calls, but I need to know what serial port to reference.

**Comment by 3d-gussner** [comment]
Hello,
Here few tips I can give you:

- Please flash the firmware always with PrusaSlice to also flash the languages on the xflash chip.
- To add "debug" lines you can find different approaches see
  - See https://github.com/prusa3d/Prusa-Firmware/blob/MK3/Firmware/Marlin_main.cpp#L505
  - See https://github.com/prusa3d/Prusa-Firmware/blob/MK3/Firmware/Marlin_main.cpp#L2506-L2510
- Try to use the [MK404 sim](https://github.com/vintagepc/MK404) during development as you don't have to flash your printer all the time.

I have a dual boot PC with Ubuntu and Windows 10.
If you have a Linux system with VSCode
1. git clone the Prusa-Firmware repository
2. run `.utils/bootstrap.py` to download all dependencies
3. run `./PF-build.sh` to compile and run it in the simulator
  a. My mostly used command `./PF-build.sh -bAuto -lALL -vMK3S.h -p1 -m2`
    i. Compiles the firmware with current `-b`uild number, `-l` ALL languages, `-v` MK3S.h variant, `-p` keep `Configuration_prusa.h` to highlight `#defined` code in VSCode and `-m2` start MK404 SIM with MMU connected.
4. You can also run the CMake build and start the MK404 SIM manually 
5. use Putty or other Terminal software to connect to the serial line. MK404 creates `/tmp/simavr-uart0` where you can connect to


Hope that helps a bit.
